File: Herencia/ejercicio 3/ej3.java
import java.time.LocalDate;
import java.time.Period;

class Persona {
    String ci, nombre, apellido, celular, sexo;
    LocalDate fechaNac;

    Persona(String ci, String nombre, String apellido, String celular, LocalDate fechaNac, String sexo) {
        this.ci = ci;
        this.nombre = nombre;
        this.apellido = apellido;
        this.celular = celular;
        this.fechaNac = fechaNac;
        this.sexo = sexo;
    }

    int edad() {
        return Period.between(fechaNac, LocalDate.now()).getYears();
    }

    void mostrar() {
        System.out.println(nombre + " " + apellido + " - CI: " + ci + ", Edad: " + edad() + " años");
    }
}

class Estudiante extends Persona {
    String ru;
    LocalDate fechaIngreso;
    int semestre;

    Estudiante(String ci, String nombre, String apellido, String celular, LocalDate fechaNac, String sexo,
               String ru, LocalDate fechaIngreso, int semestre) {
        super(ci, nombre, apellido, celular, fechaNac, sexo);
        this.ru = ru;
        this.fechaIngreso = fechaIngreso;
        this.semestre = semestre;
    }

    @Override
    void mostrar() {
        super.mostrar();
        System.out.println("RU: " + ru + ", Fecha de Ingreso: " + fechaIngreso + ", Semestre: " + semestre);
    }
}

class Docente extends Persona {
    String nit, profesion, especialidad;

    Docente(String ci, String nombre, String apellido, String celular, LocalDate fechaNac, String sexo,
            String nit, String profesion, String especialidad) {
        super(ci, nombre, apellido, celular, fechaNac, sexo);
        this.nit = nit;
        this.profesion = profesion;
        this.especialidad = especialidad;
    }

    @Override
    void mostrar() {
        super.mostrar();
        System.out.println("NIT: " + nit + ", Profesión: " + profesion + ", Especialidad: " + especialidad);
    }
}

public class Main {
    public static void main(String[] args) {
        Estudiante estudiante = new Estudiante(
            "12345678",
            "Juan",
            "Pérez",
            "78945612",
            LocalDate.of(2002, 5, 15),
            "M",
            "RU001",
            LocalDate.of(2021, 2, 10),
            5
        );

        Docente docente = new Docente(
            "87654321",
            "María",
            "Lopez",
            "74185296",
            LocalDate.of(1980, 3, 20),
            "F",
            "NIT123456",
            "Ingeniera de Sistemas",
            "Inteligencia Artificial"
        );

        System.out.println("Información del Estudiante:");
        estudiante.mostrar();

        System.out.println("\nInformación del Docente:");
        docente.mostrar();
    }
}

File: Herencia/ejercicio 3/ej3.py
from datetime import date

class Persona:
    def __init__(self, ci, nombre, apellido, celular, fecha_nac, sexo):
        self.ci = ci
        self.nombre = nombre
        self.apellido = apellido
        self.celular = celular
        self.fecha_nac = fecha_nac
        self.sexo = sexo

    def edad(self):
        hoy = date.today()
        return hoy.year - self.fecha_nac.year - ((hoy.month, hoy.day) < (self.fecha_nac.month, self.fecha_nac.day))

    def mostrar(self):
        print(f"{self.nombre} {self.apellido} - CI: {self.ci}, Edad: {self.edad()} años")


class Estudiante(Persona):
    def __init__(self, ci, nombre, apellido, celular, fecha_nac, sexo, ru, fecha_ingreso, semestre):
        super().__init__(ci, nombre, apellido, celular, fecha_nac, sexo)
        self.ru = ru
        self.fecha_ingreso = fecha_ingreso
        self.semestre = semestre

    def mostrar(self):
        super().mostrar()
        print(f"RU: {self.ru}, Fecha de ingreso: {self.fecha_ingreso}, Semestre: {self.semestre}")


class Docente(Persona):
    def __init__(self, ci, nombre, apellido, celular, fecha_nac, sexo, nit, profesion, especialidad):
        super().__init__(ci, nombre, apellido, celular, fecha_nac, sexo)
        self.nit = nit
        self.profesion = profesion
        self.especialidad = especialidad

    def mostrar(self):
        super().mostrar()
        print(f"NIT: {self.nit}, Profesión: {self.profesion}, Especialidad: {self.especialidad}")

if __name__ == "__main__":
    fecha_nac_est = date(2002, 5, 15)
    fecha_ingreso = date(2021, 2, 10)

    estudiante = Estudiante(
        ci="12345678",
        nombre="Juan",
        apellido="Pérez",
        celular="78945612",
        fecha_nac=fecha_nac_est,
        sexo="M",
        ru="RU001",
        fecha_ingreso=fecha_ingreso,
        semestre=5
    )

    fecha_nac_doc = date(1980, 3, 20)
    docente = Docente(
        ci="87654321",
        nombre="María",
        apellido="Lopez",
        celular="74185296",
        fecha_nac=fecha_nac_doc,
        sexo="F",
        nit="NIT123456",
        profesion="Ingeniera de Sistemas",
        especialidad="Inteligencia Artificial"
    )

    print("Información del Estudiante:")
    estudiante.mostrar()
    print("\nInformación del Docente:")
    docente.mostrar()
